### ISSUE TITLE
tensordot output storage owned by something else than the tensor

### ISSUE THREAD

**Issue by pablolh** [pytorch/pytorch#153542]
### 🐛 Describe the bug

The output of `torch.tensordot` has a storage that "cannot be deleted". I have this use case to free GPU memory occupied by a tensor regardless of the reference count managed by Python. This occurs when `requires_grad` is `False`.

It seems that `detach` fixes it.

```
import torch

m = torch.eye(1000, device="cuda:0")
x = torch.tensordot(m, m, dims=1)

print(torch.cuda.memory_allocated() * 1e-6)

x.resize_(0)
x.set_(source=torch.zeros(0, device="cuda:0").untyped_storage())

print(torch.cuda.memory_allocated() * 1e-6)
```

Output:
```
16.520191999999998
16.520191999999998
```

Now replace the creation of `x` with
```
x = torch.tensordot(m, m, dims=1).detach()
```
and get the output:
```
16.520191999999998
12.519936  # x's memory is freed as I expect
```

Surprisingly, replacing `tensordot` with `matmul` doesn't need `detach`:
```
x = m @ m
```
gives
```
16.520191999999998
12.519936
```

### Versions

Collecting environment information...
PyTorch version: 2.7.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Feb  4 2025, 14:57:36) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.8.0-59-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.4.131
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4060 Laptop GPU
Nvidia driver version: 550.144.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               20
On-line CPU(s) list:                  0-19
Vendor ID:                            GenuineIntel
Model name:                           13th Gen Intel(R) Core(TM) i9-13900H
CPU family:                           6
Model:                                186
Thread(s) per core:                   2
Core(s) per socket:                   14
Socket(s):                            1
Stepping:                             2
CPU max MHz:                          5400,0000
CPU min MHz:                          400,0000
BogoMIPS:                             5990.40
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            544 KiB (14 instances)
L1i cache:                            704 KiB (14 instances)
L2 cache:                             11,5 MiB (8 instances)
L3 cache:                             24 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-19
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] flake8==7.0.0
[pip3] flake8-docstrings==1.7.0
[pip3] mypy==1.2.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] pyproject-flake8==7.0.0
[pip3] torch==2.7.0
[pip3] torch-geometric==2.6.1
[pip3] triton==3.3.0
[conda] Could not collect

**Comment by Aritra-prediqt** [comment]
**Root cause**:
`tensordot` internally creates computation graph traces or views even when `requires_grad=False`. These retain references to the underlying storage, preventing memory from being immediately released — even after explicit `resize_(0)` and `set_()` operations.

In contrast, `matmul` does not create such internal references, allowing memory to be reclaimed immediately.

**Confirmed behavior**:
Calling `.detach()` removes autograd traces and allows GPU memory to be freed as expected.

**Comment by aritra3520** [comment]
**Root cause:**
tensordot internally creates computation graph traces or views even when requires_grad=False. These retain references to the underlying storage, preventing memory from being immediately released — even after explicit resize_(0) and set_() operations.

In contrast, matmul does not create such internal references, allowing memory to be reclaimed immediately.

**Confirmed behavior:**
Calling .detach() removes autograd traces and allows GPU memory to be freed as expected.

**Comment by soulitzer** [comment]
This issue is not specific to tensordot, it is because tensordot creates a view as @aritra3520 mentions. 

Although the set_ swaps out the storage, the view metadata is unchanged.

**Comment by pablolh** [comment]
Thanks for the input! 🚀 
The fact that `tensordot` returns a view makes sense given its implementation, but is surprising if you compare with `matmul`.
Is there a way to access/change this view metadata? And somehow deallocate the memory that way?
Specifically, my problem with `detach` is that I _cannot_ use it to create a "deallocation" function like so:

```
def deallocate_tensor_with_detach(t: torch.Tensor) -> None:
    t = t.detach()
    t.resize_(0)
    t.set_(source=torch.zeros(0, dtype=t.dtype, device=t.device).untyped_storage())
```

Indeed, if I call `detach` inside that function body, the outer calling scope still keeps the storage alive. For example, that _doesn't_ deallocate:
```
x = torch.tensordot(...)
deallocate_tensor_with_detach(x)
```

but that works:

```
x = torch.tensordot(...).detach()
deallocate_tensor(x)
```

**Comment by soulitzer** [comment]
You can do `x._base.set_(...)`

**Comment by pablolh** [comment]
Thanks, it works for most cases to set the base's storage as well, when there is a single view pointing to a tensor. I discovered an interesting example:
```
m = torch.rand(123, 123, dtype=torch.complex128)
x = m.conj()
print(m._use_count())  # 3 
```
Why is there a hidden view for complex conjugation? I expected the `_use_count()` to be 2

**Comment by soulitzer** [comment]
The conj operation is a view as an optimization. It just sets "conj bit" indicating that the imag component is negated. This way no actual compute needs to be done until actually necessary.

**Comment by pablolh** [comment]
I understand, but it seems that `_use_count()` is one higher than expected, no?
For instance, I cannot deallocate if there is a `.conj()` view, even when changing also the base's storage:

```
import torch

m = torch.rand(200, 200, dtype=torch.complex128, device="cuda:0")

x = m.conj()

s = torch.zeros(0, dtype=torch.complex128, device="cuda:0").untyped_storage()

print(torch.cuda.memory_allocated(0) * 1e-6)  # 0.64

# Try to deallocate memory
x.resize_(0)
x.set_(source=s)

assert x._base is m
assert m._base is None

m.resize_(0)
m.set_(source=s)

print(torch.cuda.memory_allocated(0) * 1e-6)  # 0.64
```
